src/main/java/io/cdap/plugin/http/source/batch/HttpBatchSource.java

```diff
@@ -92,55 +90,68 @@ public void prepareRun(BatchSourceContext context) {
     context.setInput(Input.of(config.referenceName, new HttpInputFormatProvider(config)));
   }
 
-  private BaseStringToRecordConverter stringToRecordConverter;
-
   @Override
   public void initialize(BatchRuntimeContext context) throws Exception {
     this.schema = config.getSchema();
-    this.stringToRecordConverter = StringToRecordConverterFactory.createInstance(config, schema);
-    this.httpErrorHandler = new HttpErrorHandler(config);
-
     super.initialize(context);
   }
 
   @Override
-  public void transform(KeyValue<IntWritable, Text> input,
-                        Emitter<StructuredRecord> emitter) throws IOException {
-    String resultString = input.getValue().toString();
-    try {
-      int httpCode = input.getKey().get();
-      HttpErrorHandlingStrategy httpRetryStrategy = httpErrorHandler.getErrorHandlingStrategy(httpCode)
-        .getAfterRetryStrategy();
-
-      if (httpRetryStrategy.equals(HttpErrorHandlingStrategy.SEND_TO_ERROR)) {
-        emitter.emitError(buildError(httpCode, resultString,
-                                     String.format("Request failed with '%d' http status code", httpCode)));
-      } else if (httpRetryStrategy.equals(HttpErrorHandlingStrategy.SKIP)) {
-        return;
-      } else {
-        StructuredRecord record = stringToRecordConverter.getRecord(resultString);
-        emitter.emit(record);
-      }
-    } catch (Exception ex) {
-      switch (config.getErrorHandling()) {
-        case SKIP:
-          LOG.warn("Cannot convert row '{}' to a record", resultString, ex);
-          break;
-        case SEND:
-          emitter.emitError(buildError(0, resultString,
-                                       ex.getClass().getName() + ": " + ex.getMessage()));
-          break;
-        case STOP:
-          throw ex;
-        default:
-          throw new UnexpectedFormatException(
-            String.format("Unknown error handling strategy '%s'", config.getErrorHandling()));
+  public void transform(KeyValue<NullWritable, BasePage> input, Emitter<StructuredRecord> emitter) {
+    BasePage page = input.getValue();
+
+    if (handleHttpError(page, emitter)) {
+      return;
+    }
+
+    while (page.hasNext()) {
```

Comment:
It's not normal for an iterator to throw exceptions.

Instead of iterating in this method, it's better if BasePage has an emitRecords() method that takes an Emitter, with the Page class responsible for reading through the response and emitting output or errors. This puts the error handling in one place rather than some here and some in the Page classes.

Reply:
I will re-use page logic for streaming where there is no emitter. That's why I don't want to pass emitter there.

Reply:
If this logic is going to be re-used, it's even more important for it to be in some common place rather than in the transform method. You don't have to pass in an emitter then, but can use something more generic like:

```
void consume(Consumer<StructuredRecord> onOutput, Consumer<InvalidEntry<StructuredRecord>> onError);
```

Reply:
Done